security/security.c

```diff
@@ -86,6 +100,67 @@ static __initconst const char * const builtin_lsm_order = CONFIG_LSM;
 static __initdata struct lsm_info **ordered_lsms;
 static __initdata struct lsm_info *exclusive;
 
+/*
+ * Necessary information about a static
+ * slot to call __static_call_update
+ */
+struct security_static_slot {
+	/* static call key as defined by STATIC_CALL_KEY */
+	struct static_call_key *key;
+	/* static call tramp as defined by STATIC_CALL_TRAMP */
+	void *tramp;
+};
+
+/* Table of the static calls for each LSM hook */
+struct security_list_static_slots {
+	#define LSM_HOOK(RET, DEFAULT, NAME, ...) \
+		struct security_static_slot NAME[SECURITY_STATIC_SLOT_COUNT];
+	#include <linux/lsm_hook_defs.h>
+	#undef LSM_HOOK
+} __randomize_layout;
+
+/*
+ * Index of the first used static call for each LSM hook
+ * in the corresponding security_list_static_slots table.
+ * All calls with greater indices are used, INT_MAX if no call is used.
+ */
+struct security_first_static_slot_idx {
+	#define LSM_HOOK(RET, DEFAULT, NAME, ...) \
+		int NAME;
+	#include <linux/lsm_hook_defs.h>
+	#undef LSM_HOOK
+} __randomize_layout;
+
+/* Create the static slots for each LSM hook */
+#define CREATE_STATIC_SLOT(NUM, NAME, RET, ...)				\
+	DEFINE_STATIC_CALL_NULL(STATIC_SLOT(NAME, NUM),			\
+				*((RET(*)(__VA_ARGS__))NULL));
+
+#define LSM_HOOK(RET, DEFAULT, NAME, ...)				\
+	SECURITY_FOREACH_STATIC_SLOT(CREATE_STATIC_SLOT, NAME, RET, __VA_ARGS__)
+#include <linux/lsm_hook_defs.h>
+#undef LSM_HOOK
+#undef CREATE_STATIC_SLOT
+
+static struct security_list_static_slots security_list_static_slots
+__initdata = {
+#define DEFINE_SLOT(NUM, NAME)						\
+	(struct security_static_slot) {					\
+		.key = &STATIC_CALL_KEY(STATIC_SLOT(NAME, NUM)),	\
+		.tramp = &STATIC_CALL_TRAMP(STATIC_SLOT(NAME, NUM))	\
+	},
+#define LSM_HOOK(RET, DEFAULT, NAME, ...)				\
+	.NAME = {							\
+		SECURITY_FOREACH_STATIC_SLOT(DEFINE_SLOT, NAME)		\
+	},
+#include <linux/lsm_hook_defs.h>
+#undef LSM_HOOK
+#undef DEFINE_SLOT
+};
+
+static struct security_first_static_slot_idx security_first_static_slot_idx
```

Comment:
I think each of these data structures you define here needs a comment, and when you use the macro LSM_HOOK trick it's worth adding in the comment on how this looks like after the macro has been preprocessed. We want this to be readable